planner/core/logical_plan_builder.go

```diff
@@ -4226,6 +4226,18 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
 		}
 	}
 
+	// check whether the path's index has a tidb_shard() prefix and the index column count more than 1
+	// e.g. index(tidb_shard(a), a)
+	for _, path := range ds.possibleAccessPaths {
+		if !path.IsTablePath() {
+			col := expression.IndexCol2Col(ds.Columns, ds.schema.Columns, path.Index.Columns[0])
+			if col != nil && expression.GcColumnExprIsTidbShard(col.VirtualExpr) && len(path.Index.Columns) > 1 {
+				path.IsShardIndexPath = true
+				ds.containExprPrefixUk = true
```

Comment:
Doesn't check the unique?

Reply:
got it

Reply:
In the `(ds *DataSource) AddPrefix4ShardIndexes` , use path.IsUkShardIndex() to check whether it is a unique shard index.
I wanted to support shard index for both non-unique and unique index at first, but we decided to support shard index only for unique index, I left it for future.
Do you think I should remove fix it?

Reply:
I fixed it